### ISSUE TITLE
llama 3.1 8B OOM with nightly build 02272025

### ISSUE THREAD

**Issue by zpcore** [AI-Hypercomputer/torchprime#127]
We notice llama 3.1 8B OOM issue in 02272025 PTXLA nightly image. The issue doesn't happen on PTXLA r2.6 release.

Model/exp config: seq_len: 8k, batch: 4 batch/device, minibatch on.

**Comment by zpcore** [comment]
We need to resolve the issue first before hostoffloading since the model is already OOMed with activation checkpointing.

**Comment by zpcore** [comment]
I have confirmed that the OOM issue is due to https://github.com/pytorch/xla/pull/8654 not the libtpu.

- Run without the PR [link](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22tpu-prod-env-one-vm%22%0Aresource.labels.location%3D%22asia-northeast1%22%0Aresource.labels.cluster_name%3D%22bodaborg-v6e-256-rxc%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22piz-xpk-v6e-256-1-20250228-192222-slice-job-0-0-%22%20severity%3E%3DDEFAULT;storageScope=project;cursorTimestamp=2025-02-28T19:48:33.631646227Z;startTime=2025-02-27T20:59:52.483Z;endTime=2025-02-28T21:29:52.483Z?e=13802955&mods=allow_workbench_image_override&project=tpu-prod-env-one-vm&inv=1&invt=AbqzcA)
- Run with the PR [link](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22tpu-prod-env-one-vm%22%0Aresource.labels.location%3D%22asia-northeast1%22%0Aresource.labels.cluster_name%3D%22bodaborg-v6e-256-rxc%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22piz-xpk-v6e-256-1-20250227-230151-slice-job-0-0-%22%20severity%3E%3DDEFAULT;storageScope=project;duration=P1D?e=13802955&mods=allow_workbench_image_override&project=tpu-prod-env-one-vm)

**Comment by zpcore** [comment]
cc @tengyifei , @bhavya01  in case you show memory increase issue.

**Comment by tengyifei** [comment]
When I access the two links, both jobs appear to succeed. They seem to also use similar HBM.

**Comment by zpcore** [comment]
> When I access the two links, both jobs appear to succeed. They seem to also use similar HBM.

Sorry I messed up with the link. Just updated with detailed run.

**Comment by tengyifei** [comment]
@zpcore could you share a `tp run` command to repro?

**Comment by tengyifei** [comment]
The linked log looks like a run-time OOM. Usually that means we ended up transferring a bunch of tensor to the TPU dynamically at runtime and that OOMed the TPU.

**Comment by zpcore** [comment]
> [@zpcore](https://github.com/zpcore) could you share a `tp run` command to repro?

```
$ tp use --project tpu-prod-env-one-vm --zone asia-northeast1 --cluster bodaborg-v6e-256-rxc --num-slices 1 --artifact-dir gs://torchprime/yifeit-exp --tpu-type v6e-256
$ LIBTPU_INIT_ARGS="--xla_tpu_scoped_vmem_limit_kib=98304 --xla_tpu_use_minor_sharding_for_major_trivial_input=true --xla_tpu_relayout_group_size_threshold_for_reduce_scatter=1 --xla_tpu_assign_all_reduce_scatter_layout=true --xla_tpu_enable_data_parallel_all_reduce_opt=true --xla_tpu_data_parallel_opt_different_sized_ops=true --xla_tpu_enable_async_collective_fusion=true --xla_tpu_enable_async_collective_fusion_fuse_all_gather=true --xla_tpu_enable_async_collective_fusion_multiple_steps=true  --xla_tpu_overlap_compute_collective_tc=true  --xla_enable_async_all_gather=true --xla_tpu_enable_async_collective_fusion_fuse_all_reduce=false  --xla_tpu_enable_sparse_core_collective_offload_all_reduce=true --xla_tpu_enable_all_reduce_offload_tracing=true --xla_tpu_use_tc_device_shape_on_sc=true  --xla_sc_enable_instruction_fusion=false  --xla_sc_disjoint_spmem=false  --xla_sc_disable_megacore_partitioning=true  --2a886c8_chip_config_name=megachip_tccontrol" XLA_IR_DEBUG=1 XLA_HLO_DEBUG=1 tp run torchprime/torch_xla_models/train.py model=llama-3.1-8b dataset_config_name=wikitext-103-raw-v1 global_batch_size=1024 profile_step=3 ici_mesh.fsdp=256
```

**Comment by tengyifei** [comment]
@zpcore I'm not sure what's going on. I ran your command with the `main` branch of torchprime, which has the Feb 27, 2025 nightly build, and cannot reproduce the OOM. I then reverted the base docker to `us-central1-docker.pkg.dev/tpu-pytorch-releases/docker/xla:r2.6.0_3.10_tpuvm_cxx11`. That also succeeded without OOMing.

## Feb 27 nightly torch_xla results

Logs: http://shortn/_Xlg5WhXlU4
Artifacts: http://shortn/_mhS6iy8JPd
HBM: 27.37 GiB
Step duration: 5.788s
Profile: http://shortn/_QjLAUnNXMP

## torch_xla 2.6 stable results

Logs: http://shortn/_BxcnsH5Dw7
Artifacts: http://shortn/_YdZ74CsZcF
HBM: 27.28 GiB
Step duration: 5.780s
Profile: http://shortn/_yfoElv9f5R

I guess there's a teeny bit of regression in HBM and MFU but it's almost negligible.
